app/src/gecko/java/org/mozilla/focus/web/WebViewProvider.java

```diff
@@ -96,34 +139,106 @@ public void cleanup() {
 
         @Override
         public void setBlockingEnabled(boolean enabled) {
-            // We can't actually do this?
+            if (enabled) {
+                updateBlocking();
+            } else {
+                if (geckoSession != null) {
+                    geckoSession.disableTrackingProtection();
+                }
+            }
+            if (callback != null) {
+                callback.onBlockingStateChanged(enabled);
+            }
         }
 
-        private ContentListener createContentListener() {
-            return new ContentListener() {
+        @Override
+        public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String prefName) {
+            if (!prefName.isEmpty() && sharedPreferences != null) {
+                if (prefName.equals(getContext().getResources().getString(R.string.pref_key_privacy_block_social))) {
+                    socialTrackersBlocked = sharedPreferences.getBoolean(prefName, true);
+                } else if (prefName.equals(getContext().getResources().getString(R.string.pref_key_privacy_block_ads))) {
+                    adTrackersBlocked = sharedPreferences.getBoolean(prefName, true);
+                } else if (prefName.equals(getContext().getResources().getString(R.string.pref_key_privacy_block_analytics))) {
+                    analyticTrackersBlocked = sharedPreferences.getBoolean(prefName, true);
+                } else if (prefName.equals(getContext().getResources().getString(R.string.pref_key_privacy_block_other))) {
+                    contentTrackersBlocked = sharedPreferences.getBoolean(prefName, false);
+                }
+            }
+            updateBlocking();
+        }
+
+        private void updateBlocking() {
+            int categories = 0;
+            if (socialTrackersBlocked) {
+                categories += GeckoSession.TrackingProtectionDelegate.CATEGORY_SOCIAL;
+            }
+            if (adTrackersBlocked) {
+                categories += GeckoSession.TrackingProtectionDelegate.CATEGORY_AD;
+            }
+            if (analyticTrackersBlocked) {
+                categories += GeckoSession.TrackingProtectionDelegate.CATEGORY_ANALYTIC;
+            }
+            if (contentTrackersBlocked) {
+                categories += GeckoSession.TrackingProtectionDelegate.CATEGORY_CONTENT;
+            }
+            if (geckoSession != null) {
+                geckoSession.enableTrackingProtection(categories);
+            }
+        }
+
+        private GeckoSession.ContentListener createContentListener() {
```

Comment:
\[PMD\] INFO: The method 'createContentListener' has a Modified Cyclomatic Complexity of 10\.